### ISSUE TITLE
Incompatibility with zmq 4.x

### ISSUE THREAD

**Issue by birkenfeld** [erickt/rust-zmq#69]
The readme doesn't state the required libzmq version, so I was trying with 4.1.1. This appears to be binary incompatible with 3.x.

running helloworld_server example:
```
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Bad address', /var/tmp/portage/dev-lang/rust-999/work/rustc-nightly/src/libcore/result.rs:731
stack backtrace:
   1:     0x7f0d24732b53 - sys::backtrace::write::hcf068903b51cfe2bkds
   2:     0x7f0d24735df1 - panicking::on_panic::h417cbf9fca890635fWw
   3:     0x7f0d2472e8ea - rt::unwind::begin_unwind_inner::haa3e65fd351aaf75ZBw
   4:     0x7f0d2472ec07 - rt::unwind::begin_unwind_fmt::h6eccac85ccefdd175Aw
   5:     0x7f0d24735746 - rust_begin_unwind
   6:     0x7f0d2475bed4 - panicking::panic_fmt::h6bb5eab6a35b4ca8i5B
   7:     0x7f0d2472bea0 - main::h3104d7f3d30ba747gaa
   8:     0x7f0d2473a428 - rust_try_inner
   9:     0x7f0d2473a415 - rust_try
  10:     0x7f0d24737593 - rt::lang_start::h3197185c34da9942jRw
  11:     0x7f0d23f18f9f - __libc_start_main
  12:     0x7f0d2472bad8 - <unknown>
  13:                0x0 - <unknown>
thread '<main>' panicked at 'assertion failed: `(left == right)` (left: `-1`, right: `0`)', src/lib.rs:597
stack backtrace:
   1:     0x7f0d24732b53 - sys::backtrace::write::hcf068903b51cfe2bkds
   2:     0x7f0d24735df1 - panicking::on_panic::h417cbf9fca890635fWw
   3:     0x7f0d2472e8ea - rt::unwind::begin_unwind_inner::haa3e65fd351aaf75ZBw
   4:     0x7f0d2472ec07 - rt::unwind::begin_unwind_fmt::h6eccac85ccefdd175Aw
   5:     0x7f0d2472d0d7 - Message.Drop::drop::h0b916bc8e49fb312h0a
   6:     0x7f0d2472bec1 - main::h3104d7f3d30ba747gaa
   7:     0x7f0d2473a428 - rust_try_inner
   8:     0x7f0d2473a415 - rust_try
   9:     0x7f0d24737593 - rt::lang_start::h3197185c34da9942jRw
  10:     0x7f0d23f18f9f - __libc_start_main
  11:     0x7f0d2472bad8 - <unknown>
  12:                0x0 - <unknown>
thread panicked while panicking. aborting.
An unknown error occurred
```
running helloworld_client example:
```
Connecting to hello world server...

Sending Hello 0
Assertion failed: check () (/var/tmp/portage/net-libs/zeromq-4.1.1/work/zeromq-4.1.1/src/msg.cpp:220)
An unknown error occurred
```

**Comment by blaenk** [comment]
Well I'm pretty sure it used to work for me with 4.0.x, but it seems that 4.1.x bumped the ABI version or something. For example this triggers a `SIGABRT` due to failing a zeromq-internal assertion:

``` rust
Message::from_slice(b"data")
```

Perhaps it's just a matter of re-running rust-bindgen? It'd be nice if we had access to a script that run it with the intended parameters, so we could do it ourselves.

**Comment by dnaeon** [comment]
Same issue here as well, running with libzmq 4.1.2

On a side note I've noticed a lot of inconsistencies between the libzmq C API and the current `zmq-sys` low-level bindings. Have a look at the error consts for example - the Rust consts simply don't match to the ones in the C library.

@blaenk I've ran `rust-bindgen` against the 4.1.2 version of libzmq and to be honest the result doesn't look as pretty as I'd like it to be. Perhaps it's better to just translate `zmq.h` manually from scratch as that would be easier to maintain in the long run.

I might just give it a try in the next days and see how it goes.
